dex/farm_with_community_rewards/src/custom_rewards.rs

```diff
@@ -0,0 +1,98 @@
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+use community_rewards::MINIMUM_REWARDING_BLOCKS;
+use contexts::generic::StorageCache;
+
+#[elrond_wasm::module]
+pub trait CustomRewardsModule:
+    config::ConfigModule
+    + token_send::TokenSendModule
+    + farm_token::FarmTokenModule
+    + rewards::RewardsModule
+    + community_rewards::CommunityRewardsModule
+    + pausable::PausableModule
+    + elrond_wasm_modules::only_admin::OnlyAdminModule
+    + elrond_wasm_modules::default_issue_callbacks::DefaultIssueCallbacksModule
+{
+    fn distribute_per_block_rewards(&self) -> BigUint {
+        let current_block_nonce = self.blockchain().get_block_nonce();
+        let last_reward_nonce = self.last_reward_block_nonce().get();
+
+        if current_block_nonce > last_reward_nonce {
+            let mut to_distribute =
+                self.calculate_per_block_community_rewards(current_block_nonce, last_reward_nonce);
+
+            if to_distribute != 0 {
+                let community_rewards_remaining_reserve_mapper =
+                    self.community_rewards_remaining_reserve();
+                let community_rewards_remaining_reserve =
+                    community_rewards_remaining_reserve_mapper.get();
+
+                if to_distribute >= community_rewards_remaining_reserve {
+                    to_distribute = community_rewards_remaining_reserve;
+                    community_rewards_remaining_reserve_mapper.clear();
+                    self.produce_community_rewards_enabled().set(false);
+                } else {
+                    community_rewards_remaining_reserve_mapper
+                        .update(|total| *total -= &to_distribute);
+                }
+            }
+            self.last_reward_block_nonce().set(&current_block_nonce);
+            to_distribute
```

Comment:
And doesn't it work as intended? First you compute `to_distribute`. Then, if it is greater than what we have remaining in the reserve, we distribuite the entire remaining reserve and clear the storage. Otherwise, we just adjust the remaining reserves storage, as `to_distribute` is already computed.

Reply:
The storage is set correctly, but the value returned (`to_distribute`) is then used to compute the rewards per share, which is increased by more than the available remaining funds. This means that the contract will run out of funds before all users can claim their share.

Reply:
I still don't see the issue. I assume you refer to the case where `to_distribute >= community_rewards_remaining_reserve`, but here, `to_distribute` variable is overwritten with what rewards are still available and the rps should be computed as intended. Also, at that point, the rewards distribution is paused and new rewards must be deposited for the distribution to be started again.
Maybe you could suggest an example with what needs to be modified, to better understand the issue.

Reply:
Ah, you're right, I missed the line `to_distribute = community_rewards_remaining_reserve;`